### PR TITLE
Update design and expose new attributes for the CustomTypeMarshaller support

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Update the design of the V2 marshallers based on API review feedback.

Expose the `CustomTypeMarshallerAttribute` class and the `MarshalMode` enum so we can use them to update ASP.NET Core's custom marshallers and delete the V1 model sooner.

Also block the analyzer from warning on the V2 design for now until we actually get it updated.

Contributes to #70859

### CODE REVIEW

File: docs/design/libraries/LibraryImportGenerator/UserTypeMarshallingV2.md [hunk 1]
```diff
@@ -448,14 +444,14 @@ We'll continue with the collection marshaller shapes. These marshaller shapes su
 Each of these shapes will support marshalling the following type:
 
 ```csharp
-// Any number of generic parameters is allowed, with any constraints
+// Any number of generic parameters is allowed, with any constraints]
```
Comment (AaronRobinsonMSFT) [inline docs/design/libraries/LibraryImportGenerator/UserTypeMarshallingV2.md:447]:
```suggestion
// Any number of generic parameters is allowed, with any constraints
```

File: docs/design/libraries/LibraryImportGenerator/UserTypeMarshallingV2.md [hunk 2]
```diff
@@ -604,7 +606,8 @@ The element type of the `Span` for the caller-allocated buffer can be any type t
 ```csharp
 [ManagedToUnmanagedMarshallers(typeof(TCollection<,,,...>), InMarshaller = typeof(ManagedToNative))]
 [UnmanagedToManagedMarshallers(typeof(TCollection<,,,...>), OutMarshaller = typeof(ManagedToNative))]
-static class TMarshaller<T, U, V..., [ElementUnmanagedType] TUnmanagedElement> where TUnmanagedElement : unmanaged
+[ContiguousCollectionMarshaller]
+static class TMarshaller<T, U, V..., TUnmanagedElement> where TUnmanagedElement : unmanaged
 {
     [CustomTypeMarshallerFeatures(BufferSize = 0x200)]
```
Comment (AaronRobinsonMSFT) [inline docs/design/libraries/LibraryImportGenerator/UserTypeMarshallingV2.md:612]: This is old, right?

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/CustomMarshallerAttribute.cs [hunk 3]
```diff
@@ -2,28 +2,32 @@
...
         /// <param name="marshallerType">Type used for marshalling.</param>
-        public CustomMarshallerAttribute(Type managedType, Scenario scenario, Type marshallerType) { }
+        public CustomMarshallerAttribute(Type managedType, MarshalMode marshalMode, Type marshallerType)
+        {
+            ManagedType = managedType;
+            MarshalMode = marshalMode;
+            MarshallerType = marshallerType;
+        }
+
+        public Type ManagedType { get; }
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/CustomMarshallerAttribute.cs:26]: Let's add doc to these properties while we are here.

File: src/libraries/System.Runtime.InteropServices/ref/System.Runtime.InteropServices.cs [hunk 4]
```diff
@@ -2169,6 +2169,19 @@ namespace System.Runtime.InteropServices.Marshalling
...
         public void FreeNative() { }
     }
+
+    [System.AttributeUsageAttribute(System.AttributeTargets.Struct | System.AttributeTargets.Class, AllowMultiple = true)]
+    public sealed partial class CustomMarshallerAttribute : System.Attribute
+    {
+        public CustomMarshallerAttribute(System.Type managedType, System.Runtime.InteropServices.Marshalling.MarshalMode marshalMode, System.Type marshallerType) { }
+        public Type ManagedType { get { throw null; } }
+        public MarshalMode MarshalMode { get { throw null; } }
+        public Type MarshallerType { get { throw null; } }
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/ref/System.Runtime.InteropServices.cs:2179]:
```suggestion
        public System.Type ManagedType { get { throw null; } }
        public MarshalMode MarshalMode { get { throw null; } }
        public System.Type MarshallerType { get { throw null; } }
```

**Review by AaronRobinsonMSFT — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/MarshalMode.cs [hunk 5]
```diff
@@ -4,18 +4,13 @@
 #if MICROSOFT_INTEROP_SOURCEGENERATION
 namespace Microsoft.Interop
 #else
-namespace System.Runtime.InteropServices
+namespace System.Runtime.InteropServices.Marshalling
 #endif
 {
     /// <summary>
     /// An enumeration representing the different marshalling scenarios in our marshalling model.
```
Comment (elinor-fung) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/MarshalMode.cs:11]:
```suggestion
    /// An enumeration representing the different marshalling modes in our marshalling model.
```

**Review by elinor-fung — APPROVED** [review]

**Comment by jkoritzinsky** [comment]
Failures are a network blip and https://github.com/dotnet/runtime/issues/71684

**Pull request merged**